sentry-core/src/main/java/io/sentry/core/EnvelopeReader.java

```diff
@@ -0,0 +1,211 @@
+package io.sentry.core;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.TypeAdapter;
+import com.google.gson.stream.JsonReader;
+import com.google.gson.stream.JsonWriter;
+import io.sentry.core.protocol.SentryId;
+import io.sentry.core.util.Nullable;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class EnvelopeReader {
+
+  private final Gson gson =
+      new GsonBuilder()
+          .registerTypeAdapter(
+              SentryEnvelopeHeader.class, new EnvelopeReader.SentryEnvelopeHeaderAdapter())
+          .registerTypeAdapter(
+              SentryEnvelopeItemHeader.class, new EnvelopeReader.SentryEnvelopeItemHeaderAdapter())
```

Comment:
Didn't want to rely on reflection here. But it's an option.
I wanted to discuss with you about using this strategy for the whole protocol.

Reply:
got it.
it could be, but I really don't see any benefit of adding a 3rd party and not relying on reflection nor code generation, we still will depend on a 3rd party and do halfway of parsing still manually, but let's discuss it separately.

Reply:
Seems like `Iterable` is API 1, only the `.forEach` on it is API 24:
https://developer.android.com/reference/java/lang/Iterable#forEach(java.util.function.Consumer%3C?%20super%20T%3E)

I'd suggest we stick with `Iterable`s then, anyway the tooling won't let us use `iterable.forEach` on it.
Agreed?

Reply:
Anyway if we change, it'll be a separate PR.
Happy to change btw, if we agree to go all the way wtih gson as a dependency.

Reply:
Just want to close this